### PR TITLE
Add a unique identifier for fibers

### PR DESCRIPTION
**Pull request by fabbing**
# Introduction
Currently, there is no obvious way to identify a fiber among others, even for the runtime itself.
There is no unique information in the fiber and when a fiber stack needs to be reallocated, the whole fiber is reallocated so a pointer to the current `struct stack_info` of the fiber is invalided.

In some cases it may be convenient to identify a specific fiber: 
- For the runtime/debugger to place a `trap barrier` in a specific fiber ([11065](https://github.com/ocaml/ocaml/pull/11065))
- For users: to use a profiler that instruments code and wouldn't be able to differentiate a function entry/exit per fiber; to log information per fiber...

# Changes
An unique `int64_t` is added to the `struct stack_info` to identify a fiber globally.
The `id` 0 is used for the first fiber, subsequent one got an increasing `id` number.

An `Effect.fiber_id` function is added for users to access this information.

# Example
```ocaml
open Printf
open Effect
open Effect.Deep

type _ t += E : int -> int t

let[@inline never][@local never] f x = x

let[@inline never] consume_stack () =
  let size = 128 in
  let allocated = 2 * 2 (* 2 spilled registers *) + 1 (* saved rbp *) in
  let count = size / allocated in
  let[@inline never] rec gobbler i =
    (* Force spilling of x0 and x1 *)
    let x0 = Sys.opaque_identity 42 in
    let x1 = Sys.opaque_identity 42 in
    let _ = f x0 in
    let _ = f x1 in
    let _ = Sys.opaque_identity x0 in
    let _ = Sys.opaque_identity x1 in
    let v = if i = 1 then 42 (* dummy *) else gobbler (i - 1) in
    v - 1 (* dummy *)
  in
  ignore (gobbler count)

let fiber domain =
  let comp () =
    printf "[%c %Ld] perform effect (E 0)\n%!" domain (fiber_id ());
    consume_stack ();
    let v = perform (E 0) in
    printf "[%c %Ld] perform returns %d\n%!" domain (fiber_id ()) v;
    v + 1
  in
  let effc (type a) (eff : a t) : ((a, 'b) continuation -> 'b) option =
    let effect_e v k =
      printf "[%c %Ld] caught effect (E %d). continuing...\n%!" domain (fiber_id ()) v;
      consume_stack ();
      let v = continue k (v + 1) in
      printf "[%c %Ld] continue returns %d\n%!" domain (fiber_id ()) v;
      v + 1
    in
    match eff with
    | E v -> Some (effect_e v)
    | e -> None
  in
  let retc v =
    consume_stack ();
    printf "[%c %Ld] done %d\n%!" 'A' (fiber_id ()) v;
    v + 1
  in
  match_with comp ()
  { retc = retc;
    exnc = (fun e -> raise e);
    effc = effc }

let () =
  printf "Hello from main, fiber = %Ld\n" (fiber_id ());
  let dA = Domain.spawn (fun _ -> fiber 'A') in
  let dB = Domain.spawn (fun _ -> fiber 'B') in
  ignore @@ Domain.join dA;
  ignore @@ Domain.join dB
```
This example gives the following output:
```
### OCaml runtime: debug mode ###
Hello from main, fiber = 0
[A 2] perform effect (E 0)
[A 1] caught effect (E 0). continuing...
[A 2] perform returns 1
[A 1] done 2
[A 1] continue returns 3
[B 4] perform effect (E 0)
[B 3] caught effect (E 0). continuing...
[B 4] perform returns 1
[A 3] done 2
[B 3] continue returns 3
```
---
@TheLortex

### CODE REVIEW

File: runtime/fiber.c [hunk 1]
```diff
@@ -583,3 +592,14 @@ CAMLprim value caml_drop_continuation (value cont)
   caml_free_stack(stk);
   return Val_unit;
 }
+
+CAMLprim value caml_fiber_id(value unit)
+{
+  struct stack_info* stack = Caml_state->current_stack;
+
+  CAMLassert(unit == Val_unit);
+  if (!stack || stack->id > INT_MAX)
```
Comment (fabbing) [inline runtime/fiber.c]: This should probably be improved: `id` is an `uintnat` to get the free overflow handling by C but conversion to OCaml `int` will overflow sooner!

File: stdlib/effect.ml [hunk 2]
```diff
@@ -72,6 +72,8 @@ module Deep = struct
   external get_callstack :
     ('a,'b) continuation -> int -> Printexc.raw_backtrace =
     "caml_get_continuation_callstack"
+
+  external id : unit -> int = "caml_fiber_id"
```
Comment (kayceesrk) [inline stdlib/effect.ml]: This function may be shared between deep and shallow handlers. Better to move it to the Effect module, rather than having it both the Effect.Deep and Effect.Shallow modules. 

It may be better to call this function `fiber_id`.

File: runtime/fiber.c [hunk 3]
```diff
@@ -183,20 +186,23 @@ alloc_size_class_stack_noexc(mlsize_t wosize, int cache_bucket,
...
   int cache_bucket = stack_cache_bucket (wosize);
-  return alloc_size_class_stack_noexc(wosize, cache_bucket, hval, hexn, heff);
+  return alloc_size_class_stack_noexc(wosize, cache_bucket, hval, hexn, heff,
+      id);
 }
 
 #ifdef NATIVE_CODE
 
 value caml_alloc_stack (value hval, value hexn, value heff) {
+  static __thread uintnat id = 1;
```
Comment (kayceesrk) [inline runtime/fiber.c]: I would prefer if the fiber id counter were a global variable which is incremented atomically. Currently, given that the id is a thread local variable, we may have two different fibers from different domains with the same id?

File: runtime/fiber.c [hunk 4]
```diff
@@ -315,11 +321,13 @@ value caml_global_data;
 CAMLprim value caml_alloc_stack(value hval, value hexn, value heff)
 {
   value* sp;
+  static __thread uintnat id = 1;
```
Comment (kayceesrk) [inline runtime/fiber.c]: Same as the other comment on thread local id.

**Review by kayceesrk — CHANGES_REQUESTED** [review]
The addition of fiber id is useful for the purposes stated in the PR. I have left some comments on the code.

File: runtime/fiber.c (hunk 1 above)
Comment (xavierleroy) [reply runtime/fiber.c]: Just replace `INT_MAX` by `Max_long` (defined in `<caml/mlvalues.h>`).

However, on 32-bit architectures, overflow will occur rather quickly (2^30 = 1 billion fibers created).  So I'd suggest we avoid repeating mistakes of the past, and use a 64-bit integer (`int64_t` in C, `int64` in OCaml) as identifier.  This has the additional benefit that there is no need to test for overflow, because it will never happen in your lifetime.

File: runtime/fiber.c (hunk 3 above)
Comment (xavierleroy) [reply runtime/fiber.c]: I agree with @kayceesrk that fiber ids should be globally unique.  However, an `_Atomic int64_t` may not be efficiently implemented on 32-bit platforms...  Maybe we don't care about efficiency in this case.

A way to have our cake and eat it too would be to have a thread-local counter, initialized to the domain number and incremented by `Max_domains`.

Comment (gasche) [reply runtime/fiber.c]: I'm wary of depending on Max_domains remaining constant, given that I'm dreaming of making the constant go away at some point. I remember you told me that *object* unique identifiers use a cool trick to avoid excessive synchronizations by reserving intervals of size 1024 and only synchronizing on a global counter at the boundary, why not use the same trick here? (Can we even reuse the code instead of duplicating the logic?)

Comment (kayceesrk) [reply runtime/fiber.c]: The performance of fiber stack allocation is not critical unlike switching between fibers (through `perform`, `continue` and `discontinue`). We measured the performance of these operations in the PLDI 21 paper (https://arxiv.org/pdf/2104.00250.pdf, Section 6.3). We see that installing new effect handlers (which includes allocating a fiber stack) is the most expensive operation. That said, in practice, I would expect switching between fibers to outweigh the number of fiber stack allocations by a large factor. Hence, I would go for the `_Atomic int64_t` if it keeps the code simple. 

Factoring out the code for reserving intervals seems like a useful addition. I'd prefer that we do it in a separate PR.

Comment (gasche) [reply runtime/fiber.c]: Sounds very reasonable to me.

**Comment by fabbing** [comment]
Thanks for all this valuable feedback!
I think I've addressed all the suggestions and updated the PR description text accordingly.

File: runtime/fiber.c [hunk 5]
```diff
@@ -132,7 +134,7 @@ Caml_inline int stack_cache_bucket (mlsize_t wosize) {
 
 static struct stack_info*
 alloc_size_class_stack_noexc(mlsize_t wosize, int cache_bucket,
```
Comment (kayceesrk) [inline runtime/fiber.c]: Nit: I prefer the argument list on the first line to be longer than the second

```c
alloc_size_class_stack_noexc(mlsize_t wosize, int cache_bucket, value hval, 
                             value hexn, value heff, int64_t id)
```

File: runtime/fiber.c [hunk 6]
```diff
@@ -183,18 +186,20 @@ alloc_size_class_stack_noexc(mlsize_t wosize, int cache_bucket,
 
 /* allocate a stack with at least "wosize" usable words of stack */
 static struct stack_info* alloc_stack_noexc(mlsize_t wosize, value hval,
-                                            value hexn, value heff)
```
Comment (kayceesrk) [inline runtime/fiber.c:186]:
```c
static struct stack_info*                                                           
alloc_stack_noexc(mlsize_t wosize, value hval, value hexn, value heff,              
                  int64_t id) 
```

File: runtime/fiber.c [hunk 7]
```diff
@@ -183,18 +186,20 @@ alloc_size_class_stack_noexc(mlsize_t wosize, int cache_bucket,
...
+                                            value hexn, value heff, int64_t id)
 {
   int cache_bucket = stack_cache_bucket (wosize);
-  return alloc_size_class_stack_noexc(wosize, cache_bucket, hval, hexn, heff);
```
Comment (kayceesrk) [inline runtime/fiber.c:189]:
```c
  return alloc_size_class_stack_noexc(wosize, cache_bucket, hval, hexn, heff,    
                                      id);
```

**Review by kayceesrk — APPROVED** [review]
Looks good to me. I've left some minor formatting comments.

File: runtime/fiber.c (hunk 3 above)
Comment (avsm) [reply runtime/fiber.c]: Also confirmed the same patterns post-PLDI paper in #10875, where the switch to mmap instead of the (slab allocated) malloc for fibre stack allocation made no practical performance difference.

File: runtime/fiber.c [hunk 8]
```diff
@@ -583,3 +591,14 @@ CAMLprim value caml_drop_continuation (value cont)
...
   return Val_unit;
 }
+
+CAMLprim value caml_fiber_id(value unit)
+{
+  struct stack_info* stack = Caml_state->current_stack;
+
+  CAMLassert(unit == Val_unit);
+  if (!stack)
+    return caml_copy_int64(-1);
```
Comment (avsm) [inline runtime/fiber.c]: what are the circumstances under which there's no fibre stack, and we return -1?  Shouldn't retrieving a stackless fibre result in an exception or an assertion failure?

File: runtime/fiber.c (hunk 3 above)
Comment (gasche) [reply runtime/fiber.c]: The "local reservation pattern" is a bit invasive, as it requires per-domain value (of course!), and thus more ceremony in Caml_state; the ceremony effectively discouraged me from writing a follow-up PR. I'm convinced by the idea that the memory allocation that follows dominates the performance cost of the atomic increment here, and we can always revisit if we learn otherwise.

File: runtime/fiber.c (hunk 8 above)
Comment (kayceesrk) [reply runtime/fiber.c]: Agree. There will always be a fiber when executing OCaml code. The check can be converted to an assertion.

File: runtime/caml/camlatomic.h [hunk 9]
```diff
@@ -29,8 +29,9 @@
 extern "C++" {
 #include <atomic>
 #define ATOMIC_UINTNAT_INIT(x) (x)
-typedef std::atomic<uintnat> atomic_uintnat;
-typedef std::atomic<intnat> atomic_intnat;
+using atomic_uintnat = std::atomic<uintnat>;
+using atomic_intant = std::atomic<intnat>;
+using atomic_int64 = std::atomic<int64>;
```
Comment (xavierleroy) [inline runtime/caml/camlatomic.h]: Is this valid C++?  

At any rate I don't think we need to define `atomic_int64` in a public header.  Just use `_Atomic int64` where needed.

File: runtime/fiber.c [hunk 10]
```diff
@@ -45,6 +45,8 @@
 #define fiber_debug_log(...)
 #endif
 
+static atomic_int64 fiber_id = 0;
```
Comment (xavierleroy) [inline runtime/fiber.c]:
```suggestion
static int64_t _Atomic fiber_id = 0;
```

Comment (xavierleroy) [reply runtime/fiber.c]: (Justification: we already use the `_Atomic` modifier in lf_skiplist.c for example, and I think we gave up on having enough defines and typedefs in camlatomic.h to work around C compilers that don't support `_Atomic`.)

File: runtime/fiber.c [hunk 11]
```diff
@@ -583,3 +592,14 @@ CAMLprim value caml_drop_continuation (value cont)
   caml_free_stack(stk);
   return Val_unit;
 }
+
+CAMLprim value caml_fiber_id(value unit)
+{
+  struct stack_info* stack = Caml_state->current_stack;
+
+  CAMLassert(unit == Val_unit);
```
Comment (xavierleroy) [inline runtime/fiber.c]: We don't do that (asserting that unit parameters are equal to `Val_unit`) in other parts of the runtime system.

File: stdlib/effect.mli [hunk 12]
```diff
@@ -24,6 +24,9 @@ external perform : 'a t -> 'a = "%perform"
 
     @raise Unhandled if there is no active handler. *)
 
+external fiber_id : unit -> int64 = "caml_fiber_id"
+(** [fiber_id] returns an unique identifier for the current fiber. *)
```
Comment (xavierleroy) [inline stdlib/effect.mli]:
```suggestion
(** [fiber_id] returns a unique identifier for the current fiber. *)
```
("unique" is pronounced as if it started with "you", so "an" would sound wrong there.)

**Review by xavierleroy — COMMENTED** [review]
Looks good, thanks, but see below for minor remarks.

File: runtime/fiber.c (hunk 8 above)
Comment (fabbing) [reply runtime/fiber.c]: Indeed, this shouldn't happen! I updated the code to rather use an assertion.

File: runtime/caml/camlatomic.h (hunk 9 above)
Comment (fabbing) [reply runtime/caml/camlatomic.h]: It's the "modern" typedef in C++ that also allows usage of template ([type_alias](https://en.cppreference.com/w/cpp/language/type_alias)).

Comment (xavierleroy) [reply runtime/caml/camlatomic.h]: Oh, I guess I need to brush on my C++ :-)

File: stdlib/effect.mli [hunk 13]
```diff
@@ -24,6 +24,9 @@ external perform : 'a t -> 'a = "%perform"
 
     @raise Unhandled if there is no active handler. *)
 
+external fiber_id : unit -> int64 = "caml_fiber_id"
```
Comment (avsm) [inline stdlib/effect.mli]: Do we need to expose the int64 to the external interface? 

```
module Id : sig
  type t
  val compare : t -> t -> int
  val pp : ...
end
val fiber_id : unit -> Id.t
```

would be sufficient to be able to use them now and modify the internal representation (e.g. to an unboxed "Int63" on 64-bit platforms) in the future.

**Comment by sadiqj** [comment]
Little bit late to the reviews but apart from cosmetic changes I think this looks good. The way it's implemented it should also be possible to pluck this out when doing sampling profiling or tracing.

File: stdlib/effect.mli (hunk 13 above)
Comment (fabbing) [reply stdlib/effect.mli]: If I understood correctly one advantage of exposing `Domain.id` ([domain.mli#L32](https://github.com/ocaml/ocaml/blob/trunk/stdlib/domain.mli#L32)) as a `private int` was ease of use with Array/Hashtable. On the other hand, it would be nice to avoid some allocation.
Is there already an example of such use in the `stdlib`?

Comment (xavierleroy) [reply stdlib/effect.mli]: It's safe to say that we don't know what a good API for unique identifiers looks like.  

In an attempt to defuse the issue, I'd like to point out that we do not need to expose fiber IDs to OCaml code in 5.0.  Right now the only potential use is for the bytecode debugger, which will access these IDs through a completely different, RPC-based API.  So, why don't we focus on the C implementation of fiber IDs and on the debugger interface first?

Comment (avsm) [reply stdlib/effect.mli]: Agreed that we don't need to expose this to OCaml in the 5.0 case.

Comment (fabbing) [reply stdlib/effect.mli]: > In an attempt to defuse the issue, I'd like to point out that we do not need to expose fiber IDs to OCaml code in 5.0. Right now the only potential use is for the bytecode debugger, which will access these IDs through a completely different, RPC-based API. So, why don't we focus on the C implementation of fiber IDs and on the debugger interface first?

I agree that allowing the debugger PR to complete is the most important thing.

I don't even think it's necessary to expose these fiber ids to `ocamldebug`: `trap barriers` are only used to `finish` the current function, which should be the outermost one in the last active fiber. Only the runtime needs to be able to identify the exact fiber, so it shouldn't be necessary to communicate the id to `ocamldebug`. Or did I miss something important with `checkpoint` and `trap barrier`?

**Comment by fabbing** [comment]
Rebased into a single commit with only the C implementation, no longer exposing a public API until we find a consensus!

(Branch history still available [here](https://github.com/fabbing/ocaml/tree/fiber_id_orig).)

**Review by kayceesrk — APPROVED** [review]
Looks good to me.

**Comment by sadiqj** [comment]
Thanks @fabbing !

**Comment by sadiqj** [comment]
trunk CI failed with

```List of failed tests:
    tests/memory-model/'forbidden.ml' with 1 (native) Timeout expired, killing all child processes
```

after I merged this. It doesn't look like it's related though.

File: runtime/caml/config.h [hunk 14]
```diff
@@ -209,7 +209,7 @@ typedef uint64_t uintnat;
 
 /* Number of words used in the control structure at the start of a stack
    (see fiber.h) */
-#define Stack_ctx_words 6
+#define Stack_ctx_words 7
```
Comment (xavierleroy) [inline runtime/caml/config.h:212]: It is 8 words on a 32-bit host...

**Review by xavierleroy — COMMENTED** [review]
This PR breaks the 32-bit builds:
```
runtime/caml/misc.h:213:21: error: size of array 'static_assertion_failure_line_65' is negative
     CAMLextern char static_assertion_failure_line_##l[(b) ? 1 : -1] \
                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
runtime/caml/misc.h:216:36: note: in expansion of macro 'CAML_STATIC_ASSERT_3'
 #define CAML_STATIC_ASSERT_2(b, l) CAML_STATIC_ASSERT_3(b, l)
                                    ^~~~~~~~~~~~~~~~~~~~
runtime/caml/misc.h:217:31: note: in expansion of macro 'CAML_STATIC_ASSERT_2'
 #define CAML_STATIC_ASSERT(b) CAML_STATIC_ASSERT_2(b, __LINE__)
                               ^~~~~~~~~~~~~~~~~~~~
runtime/caml/fiber.h:64:1: note: in expansion of macro 'CAML_STATIC_ASSERT'
 CAML_STATIC_ASSERT(sizeof(struct stack_info) ==
 ^~~~~~~~~~~~~~~~~~
```

**Comment by shindere** [comment]
Xavier Leroy (2022/06/01 05:22 -0700):
> @xavierleroy commented on this pull request.
>
> This PR breaks the 32-bit builts:
> ```
> runtime/caml/misc.h:213:21: error: size of array 'static_assertion_failure_line_65' is negative
>      CAMLextern char static_assertion_failure_line_##l[(b) ? 1 : -1] \
>                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> runtime/caml/misc.h:216:36: note: in expansion of macro 'CAML_STATIC_ASSERT_3'
>  #define CAML_STATIC_ASSERT_2(b, l) CAML_STATIC_ASSERT_3(b, l)
>                                     ^~~~~~~~~~~~~~~~~~~~
> runtime/caml/misc.h:217:31: note: in expansion of macro 'CAML_STATIC_ASSERT_2'
>  #define CAML_STATIC_ASSERT(b) CAML_STATIC_ASSERT_2(b, __LINE__)
>                                ^~~~~~~~~~~~~~~~~~~~
> runtime/caml/fiber.h:64:1: note: in expansion of macro 'CAML_STATIC_ASSERT'
>  CAML_STATIC_ASSERT(sizeof(struct stack_info) ==
>  ^~~~~~~~~~~~~~~~~~
> ```
>
>
> > @@ -209,7 +209,7 @@ typedef uint64_t uintnat;
>
>  /* Number of words used in the control structure at the start of a stack
>     (see fiber.h) */
> -#define Stack_ctx_words 6
> +#define Stack_ctx_words 7
>
> It is 8 words on a 32-bit host...

Probably worth defining in `s.h.in` then so that it can beproperly
adjusted at configure time.

>

File: runtime/caml/config.h (hunk 14 above)
Comment (xavierleroy) [reply runtime/caml/config.h:212]: ~~Actually, I wonder if `id` should be `_Atomic int64_t`, in which case it can use more than 8 bytes (if the compiler chooses to implement it with a lock), and then `Stack_ctx_words` cannot be a simple constant like 7 or 8....~~  No, sorry, `id` doesn't need to be atomic.  However, my question remains:

What is the purpose of `Stack_ctx_words`, anyway?

Comment (fabbing) [reply runtime/caml/config.h:212]: Sorry about that!
The following diff should fix the issue. Should I open a new PR?
```diff
--- a/runtime/caml/config.h
+++ b/runtime/caml/config.h
@@ -209,7 +209,11 @@ typedef uint64_t uintnat;
 
 /* Number of words used in the control structure at the start of a stack
    (see fiber.h) */
-#define Stack_ctx_words 7
+#ifdef ARCH_SIXTYFOUR
+#define Stack_ctx_words (6 + 1)
+#else
+#define Stack_ctx_words (6 + 2)
+#endif
```

`Stack_ctx_words` is used to initialize `Domainstate.stack_ctx_words` which in turn is used to check for stack overflow:
https://github.com/ocaml/ocaml/blob/e12b508876065723ed5fc35c0945030c9b7cd100/asmcomp/amd64/emit.mlp#L909

**Comment by shindere** [comment]
Fabrice (2022/06/01 08:04 -0700):
> @fabbing commented on this pull request.
>
>
>
> > @@ -209,7 +209,7 @@ typedef uint64_t uintnat;
>
>  /* Number of words used in the control structure at the start of a stack
>     (see fiber.h) */
> -#define Stack_ctx_words 6
> +#define Stack_ctx_words 7
>
> Sorry about that!
> The following diff should fix the issue. Should I open a new PR?
> ```diff
> --- a/runtime/caml/config.h
> +++ b/runtime/caml/config.h
> @@ -209,7 +209,11 @@ typedef uint64_t uintnat;
>
>  /* Number of words used in the control structure at the start of a stack
>     (see fiber.h) */
> -#define Stack_ctx_words 7
> +#ifdef ARCH_SIXTYFOUR
> +#define Stack_ctx_words (6 + 1)
> +#else
> +#define Stack_ctx_words (6 + 2)
> +#endif
> ```

I don't htink you need a new PR but, as I explained, it would be better
to compute this at configure time rather than to add ifdefs to the code,
IMO. Let me know if you need help to achieve this.

File: runtime/caml/config.h (hunk 14 above)
Comment (xavierleroy) [reply runtime/caml/config.h:212]: Well, if you make a branch with this code I could test it on the Jenkins CI, which includes 32-bit platforms.  And if you have a branch, a PR is not much more work.

**Comment by xavierleroy** [comment]
>  it would be better to compute this at configure time rather than to add ifdefs to the code,

It's not at all obvious to me that it's better.  Let's fix the issue first, without changing the structure of the code.

**Comment by shindere** [comment]
Well, ifdefs in the code will need to be evaluated several times. Also, if something is known at configure time why not go as far as possible in the computations, which will thus be done only once? But, it's okay to test the change first. I can even move the change to the configure script once it has been verified that it's working.

**Comment by fabbing** [comment]
@xavierleroy I've a (hopefully) better fix on this [branch](https://github.com/fabbing/ocaml/tree/fiber_id_fix)

I don't use a straightforward `sizeof(int64_t) / sizeof(value)` to compute the size in words of `int64_t id ` because `Stack_ctx_words` is used to seed `Domainstate.stack_ctx_words` using the C preprocessor and so `sizeof` are not evaluated which makes `domainstate.ml` invalid.

@shindere For this very specific case, it maybe be useful to know `sizeof(value)` at configure time but I'm not sure if this is possible.

**Comment by xavierleroy** [comment]
> @xavierleroy I've a (hopefully) better fix on this [branch](https://github.com/fabbing/ocaml/tree/fiber_id_fix)

I ran it through "CI preckeck" and the error is still there: on 32-bit hosts, `Stack_ctx_words` doesn't match `sizeof(struct stack_info)`, probably because of padding inserted to align the 64-bit field...

It is not a good idea to try and predict the size of a C struct.  Is there *any* other way to initialize `Domainstate.stack_ctx_words`?  Including @shindere's configure magic that would evaluate `sizeof(struct stack_info)` at configure-time and put it in the right place.

**Comment by xavierleroy** [comment]
Answering my own question: `AC_CHECK_SIZEOF` is pretty close to what we want, however we need a way to define `CAML_INTERNALS` while running this macro, so that the definition of `struct stack_info` is visible.

For a quick and dirty fix that will not delay 5.0 any further: consider putting the `int64_t id` *first* in the struct.  Then, no padding should be included, and the size should be (1 + 6) or (2 + 6) as you expect.

**Comment by fabbing** [comment]
I've improved the fix and it finally compiles on a 32bits OS, it's on the same [branch](https://github.com/fabbing/ocaml/tree/fiber_id_fix): I reordered `id` to the very end of the `struct stack_info` which gives free 64bits alignment with the 6 previous words.

I also tried to remove this `Stack_ctx_words` define completely:
Removed the dependency of `Domainstate.stack_ctx_words` on `Stack_ctx_words` with a utility function returning the actual `sizeof(struct stack_info)`.
```diff
--- a/utils/domainstate.ml.c
+++ b/utils/domainstate.ml.c
@@ -16,7 +16,9 @@
 
 #define CAML_CONFIG_H_NO_TYPEDEFS
 #include "config.h"
-let stack_ctx_words = Stack_ctx_words
+external stack_info_words : unit -> int = "caml_stack_info_words" [@@noalloc]
+
+let stack_ctx_words = stack_info_words ()
 
 type t =
 #define DOMAIN_STATE(type, name) | Domain_##name
 ```
 But `Stack_ctx_words` is also needed at https://github.com/ocaml/ocaml/blob/e12b508876065723ed5fc35c0945030c9b7cd100/runtime/amd64.S#L965 and I wasn't able to replace it.
